client/session.go

```diff
@@ -563,3 +595,936 @@ func (s *session) Close() error {
 	close(s.topoMapCh)
 	return s.topo.Close()
 }
+
+func (s *session) FetchBootstrapBlocksFromPeers(
+	shard uint32,
+	start, end time.Time,
+	opts bootstrap.Options,
+) (bootstrap.ShardResult, error) {
+	var (
+		result = newBlocksResult(s.opts, opts)
+		doneCh = make(chan error, 1)
+		onDone = func(err error) {
+			select {
+			case doneCh <- err:
+			default:
+			}
+		}
+		waitDone = func() error {
+			return <-doneCh
+		}
+	)
+
+	s.RLock()
+	peers := make([]hostQueue, 0, s.topoMap.Replicas())
+	err := s.topoMap.RouteShardForEach(shard, func(idx int, host topology.Host) {
+		if s.origin != nil && s.origin.ID() == host.ID() {
+			// Don't include the origin host
+			return
+		}
+		peers = append(peers, s.queues[idx])
+	})
+	s.RUnlock()
+	if err != nil {
+		return nil, err
+	}
+
+	// Begin pulling metadata, if one or multiple peers fail no error will
+	// be returned from this routine as long as one peer succeeds completely
+	metadataCh := make(chan blocksMetadata, 4096)
+	go func() {
+		err := s.streamBlocksMetadataFromPeers(shard, peers, metadataCh)
+		close(metadataCh)
+		if err != nil {
+			// Bail early
+			onDone(err)
+		}
+	}()
+
+	// Begin consuming metadata and making requests
+	go func() {
+		err := s.streamBlocksFromPeers(shard, peers, metadataCh, result)
+		onDone(err)
+	}()
+
+	if err := waitDone(); err != nil {
+		return nil, err
+	}
+	return result.result, nil
+}
+
+func (s *session) streamBlocksMetadataFromPeers(
+	shard uint32,
+	peers []hostQueue,
+	ch chan<- blocksMetadata,
+) error {
+	var (
+		wg       sync.WaitGroup
+		errLock  sync.RWMutex
+		errLen   int
+		multiErr = xerrors.NewMultiError()
+	)
+	for _, peer := range peers {
+		peer := peer
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := s.streamBlocksMetadataFromPeer(shard, peer, ch)
+			if err != nil {
+				errLock.Lock()
+				defer errLock.Unlock()
+				errLen++
+				multiErr.Add(err)
+				s.log.Warnf("failed to stream blocks metadata from peer %v for shard %d: %v", peer.Host().String(), shard, err)
+			}
+		}()
+	}
+
+	wg.Wait()
+
+	errLock.RLock()
```

Comment:
Yeah, can do, wanted to avoid data race complaining but probably is smart enough to realize this although not 100% sure.